src/main/java/org/neo4j/shell/tools/imp/format/kryo/KryoReader.java

```diff
@@ -0,0 +1,142 @@
+            while (!KryoSerializationTypes.DUMP_END.equals(type)) {
+                if (KryoSerializationTypes.NODE.equals(type)) {
+                    // Read Node
+                    long id = kryo.readObject(input, Long.class);
+                    List<String> labelStrings = kryo.readObject(input, ArrayList.class);
+                    List<Label> labels = new ArrayList<>(labelStrings.size());
+                    for (String labelString : labelStrings) {
+                        labels.add(DynamicLabel.label(labelString));
+                    }
+                    Map<String, Object> properties = kryo.readObject(input, HashMap.class);
+
+                    // Add Node to Database
+                    Node node = gdb.createNode(labels.toArray(new Label[labels.size()]));
+                    node.setProperty("id", id);
+                    if (properties != null) {
+                        for (Map.Entry<String, Object> entry : properties.entrySet()) {
+                            node.setProperty(entry.getKey(), entry.getValue());
+                        }
+                    }
+
+                    // Cache Node Id
+                    cache.put(id, node.getId());
+
+                    // Increment Transaction for Node
+                    tx.increment();
+                    reporter.update(1, 0, properties.size());
+                } else if (KryoSerializationTypes.RELATIONSHIP.equals(type)) {
+                    // Read Relationship
+                    long id = kryo.readObject(input, Long.class);
+                    long startNodeId = kryo.readObject(input, Long.class);
+                    long targetNodeId = kryo.readObject(input, Long.class);
+                    String relationshipLabel = kryo.readObject(input, String.class);
+                    Map<String, Object> properties = kryo.readObject(input, HashMap.class);
+
+                    // Add Relationship to Database
+                    RelationshipType relationshipType = relationshipLabel != null ?
+                        DynamicRelationshipType.withName(relationshipLabel) : this.relType;
+                    Node from = gdb.getNodeById(cache.get(startNodeId));
+                    Node to = gdb.getNodeById(cache.get(targetNodeId));
+                    Relationship relationship = from.createRelationshipTo(to, relationshipType);
+                    relationship.setProperty("id", id);
+                    if (properties != null) {
+                        for (Map.Entry<String, Object> entry : properties.entrySet()) {
+                            relationship.setProperty(entry.getKey(), entry.getValue());
+                        }
+                    }
+
+                    // Increment Transaction for Edge
+                    tx.increment();
+                    reporter.update(0, 1, properties.size());
+                } else {
+                    throw new RuntimeException("Encountered corrupt serialization");
```

Comment:
The only scenario where we could continue, is if there was a superfluous unknown TYPE in the file followed immediately by another TYPE.  I can make this a log statement, but I'm not sure this is a recoverable scenario in most circumstances.